File: supabase/functions/_shared/ultimate_champion_strategy.ts
import { Card, Game, PrivatePlayer } from './types.ts';
import { BotStrategy, LegalMove } from './bot_interfaces.ts';
import { canCover } from './common_utils.ts';

/**
 * Ultimate Champion Durak Strategy - Probabilistic Evolution
 * =========================================================
 * 
 * Based on advanced probabilistic analysis of 100,000 games,
 * this strategy achieved 46.1% win rate - a revolutionary improvement.
 * 
 * Combines proven boolean features with optimized probability parameters:
 * - Hard-coded proven features (trump_count_awareness, prefer_passing, etc.)
 * - 10 fine-tuned probability parameters for intelligent decision making
 * - Opponent-aware adaptive strategy
 * - Conservative give-up philosophy
 * - Sustained pressure tactics
 * 
 * This represents the pinnacle of Durak AI strategy evolution.
 */
export class UltimateChampionStrategy implements BotStrategy {
    readonly name = 'ultimate_champion';
    
    // Hard-coded proven features (from 500K game analysis)
    private readonly trump_count_awareness = true;      // +1.1% effect
    private readonly prefer_passing = true;             // +0.9% effect
    private readonly coverage_consideration = false;    // -2.2% effect
    private readonly duplicate_preference = false;      // -1.7% effect
    private readonly positional_strategy = false;       // -1.7% effect
    
    // Optimized probability parameters (from top 10 performers analysis)
    private readonly bluff_attack_prob = 0.25;           // Moderate bluffing
    private readonly give_up_threshold = 0.95;           // Very conservative about giving up
    private readonly trump_conservation_level = 0.60;    // Moderate trump conservation
    private readonly risk_taking_propensity = 0.40;      // Balanced risk-taking
    private readonly opponent_hand_weight = 0.90;        // Heavily considers opponent strength
    private readonly deck_size_sensitivity = 0.65;       // Moderately deck-aware
    private readonly attack_continuation_prob = 0.75;    // Prefers sustained pressure
    private readonly defense_desperation_threshold = 0.45;  // Moderate desperation level
    private readonly passing_aggressiveness = 0.25;      // Low-moderate passing frequency
    private readonly endgame_strategy_switch = 0.50;     // Moderate endgame adjustment
    
    async chooseMove(game: Game, botPlayerId: string, legalMoves: LegalMove[]): Promise<LegalMove> {
        if (legalMoves.length === 0) {
            throw new Error('No legal moves available');
        }
        
        const bot = game.players.find(p => p.player_id === botPlayerId) as PrivatePlayer;
        const botIndex = game.players.indexOf(bot);
        const isDefender = botIndex === game.defender;
        const isAttacker = botIndex === game.first_attacker || !isDefender;
        
        // Attack strategy
        const attackMoves = legalMoves.filter(move => move.type === 'attack');
        if (attackMoves.length > 0 && isAttacker) {
            return this.selectAttackMove(attackMoves, game, bot);
        }
        
        // Defense strategy
        const coverMoves = legalMoves.filter(move => move.type === 'cover');
        if (coverMoves.length > 0 && isDefender) {
            return this.selectDefenseMove(legalMoves, game, bot);
        }
        
        // Pass strategy (HARD-CODED PROVEN FEATURE: Prefer Passing +0.9% effect)
        if (this.prefer_passing) {
            const passMoves = legalMoves.filter(move => move.type === 'pass');
            if (passMoves.length > 0) {
                // ADVANCED FEATURE: Passing Aggressiveness
                const passChance = 0.3 + (this.passing_aggressiveness * 0.4);
                if (Math.random() < passChance) {
                    const nextDefenderIndex = (game.defender + 1) % game.players.length;
                    const nextDefender = game.players[nextDefenderIndex];
                    if (nextDefender && nextDefender.hand.length <= bot.hand.length) {
                        return this.selectPassMove(passMoves, game.power_suit);
                    }
                }
            }
        }
        
        // Good moves
        const goodMoves = legalMoves.filter(move => move.type === 'good');
        if (goodMoves.length > 0) {
            return goodMoves[0];
        }
        
        // Pickup as last resort
        const pickupMoves = legalMoves.filter(move => move.type === 'pickup');
        if (pickupMoves.length > 0) {
            return pickupMoves[0];
        }
        
        // Default: return first available move
        return legalMoves[0];
    }
    
    private selectAttackMove(attackMoves: LegalMove[], game: Game, bot: PrivatePlayer): LegalMove {
        const defenderIndex = game.defender;
        const defender = game.players[defenderIndex];
        const defenderHandSize = defender.hand.length;
        
        // ADVANCED FEATURE: Attack Continuation Probability
        if (game.table_battles.length >= 1) {
            let continueProb = this.attack_continuation_prob;
            
            // ADVANCED FEATURE: Opponent Hand Size Weight
            if (this.opponent_hand_weight > 0.5) {
                if (defenderHandSize <= 2) {
                    continueProb *= 1.5;  // More pressure against weak opponents
                } else if (defenderHandSize >= 5) {
                    continueProb *= 0.7;  // Less pressure against strong opponents
                }
            }
            
            if (Math.random() > continueProb) {
                const doneAttackMoves = attackMoves.filter(move => move.done_attacking_this_round === true);
                if (doneAttackMoves.length > 0) {
                    return doneAttackMoves[0];
                }
            }
        }
        
        // ADVANCED FEATURE: Bluff Attack Probability
        if (Math.random() < this.bluff_attack_prob) {
            const higherAttacks = attackMoves.filter(move => 
                move.cards && move.cards.some(card => card.value >= 10)
            );
            if (higherAttacks.length > 0) {
                return higherAttacks[Math.floor(Math.random() * higherAttacks.length)];
            }
        }
        
        // ADVANCED FEATURE: Trump Conservation Level
        if (this.trump_conservation_level > 0.5) {
            const nonTrumpAttacks = attackMoves.filter(move => 
                move.cards && move.cards.every(card => card.suit !== game.power_suit)
            );
            if (nonTrumpAttacks.length > 0) {
                return this.getBestAttackMove(nonTrumpAttacks, game.power_suit);
            }
        }
        
        // ADVANCED FEATURE: Risk Taking Propensity
        if (this.risk_taking_propensity > 0.7) {
            const mediumAttacks = attackMoves.filter(move => 
                move.cards && move.cards.some(card => card.value >= 8 && card.value <= 11)
            );
            if (mediumAttacks.length > 0) {
                return mediumAttacks[Math.floor(Math.random() * mediumAttacks.length)];
            }
        } else if (this.risk_taking_propensity < 0.3) {
            const lowAttacks = attackMoves.filter(move => 
                move.cards && move.cards.every(card => card.value <= 8)
            );
            if (lowAttacks.length > 0) {
                return this.getBestAttackMove(lowAttacks, game.power_suit);
            }
        }
        
        // Default: smart attack preference
        return this.getBestAttackMove(attackMoves, game.power_suit);
    }
    
    private selectDefenseMove(legalMoves: LegalMove[], game: Game, bot: PrivatePlayer): LegalMove {
        const uncoveredAttacks = game.table_battles.filter(battle => battle.defense === null);
        
        // REVOLUTIONARY FEATURE: Advanced Give-Up Decision
        if (uncoveredAttacks.length > 0) {
            const shouldGiveUp = this.ultimateGiveUpDecision(bot.hand, uncoveredAttacks, game);
            if (shouldGiveUp) {
                const pickupMoves = legalMoves.filter(move => move.type === 'pickup');
                if (pickupMoves.length > 0) {
                    console.log(`Bot ${bot.name} (ultimate_champion) chooses strategic pickup`);
                    return pickupMoves[0];
                }
            }
        }
        
        // Defense with ultimate features
        const coverMoves = legalMoves.filter(move => move.type === 'cover');
        if (coverMoves.length > 0) {
            return this.selectDefenseCard(coverMoves, game, bot);
        }
        
        // Default fallback
        return legalMoves[0];
    }
    
    private ultimateGiveUpDecision(hand: Card[], uncoveredAttacks: any[], game: Game): boolean {
        // Start with base logic
        const baseShould = this.shouldGiveUpBasic(hand, uncoveredAttacks, game);
        
        // REVOLUTIONARY FEATURE: Give Up Threshold (0.95)
        // This is the most important parameter - be very conservative about giving up
        if (Math.random() < (1.0 - this.give_up_threshold)) {
            return true;
        }
        
        // ADVANCED FEATURE: Deck Size Sensitivity
        if (this.deck_size_sensitivity > 0.5) {
            const deckRatio = game.deck.length / 36;  // Normalize to 0-1
            if (deckRatio > 0.5) {  // Lots of cards left
                if (!baseShould) {
                    return false;
                }
            } else {  // Few cards left
                if (baseShould) {
                    return true;
                }
            }
        }
        
        // ADVANCED FEATURE: Endgame Strategy Switch
        if (hand.length <= 3 && this.endgame_strategy_switch > 0.5) {
            const trumpCards = hand.filter(card => card.suit === game.power_suit);
            if (trumpCards.length >= 2) {
                return false;
            }
        }
        
        return baseShould;
    }
    
    private selectDefenseCard(coverMoves: LegalMove[], game: Game, bot: PrivatePlayer): LegalMove {
        const hand = bot.hand;
        
        // HARD-CODED PROVEN FEATURE: Trump Count Awareness (+1.1% effect)
        if (this.trump_count_awareness) {
            const trumpCount = hand.filter(card => card.suit === game.power_suit).length;
            
            // ADVANCED FEATURE: Trump Conservation Level modifies trump usage
            const conservationMultiplier = 1.0 + this.trump_conservation_level;
            
            // Adjust thresholds based on conservation level
            const manyTrumpsThreshold = Math.floor(4 * conservationMultiplier);
            const fewTrumpsThreshold = Math.max(1, Math.floor(2 / conservationMultiplier));
            
            if (trumpCount >= manyTrumpsThreshold) {
                const trumpDefenses = coverMoves.filter(move => 
                    move.cards && move.cards.some(card => card.suit === game.power_suit)
                );
                if (trumpDefenses.length > 0) {
                    return this.getLowestCostDefense(trumpDefenses, game.power_suit);
                }
            }
            
            if (trumpCount <= fewTrumpsThreshold) {
                const nonTrumpDefenses = coverMoves.filter(move => 
                    move.cards && move.cards.every(card => card.suit !== game.power_suit)
                );
                if (nonTrumpDefenses.length > 0) {
                    return this.getLowestCostDefense(nonTrumpDefenses, game.power_suit);
                }
            }
        }
        
        // ADVANCED FEATURE: Defense Desperation Threshold
        if (hand.length <= 3) {  // Low hand size
            const desperationLevel = 1.0 - this.defense_desperation_threshold;
            if (Math.random() < desperationLevel) {
                const highValueDefenses = coverMoves.filter(move => 
                    move.cards && move.cards.some(card => card.value >= 10)
                );
                if (highValueDefenses.length > 0) {
                    return this.getLowestCostDefense(highValueDefenses, game.power_suit);
                }
            }
        }
        
        // ADVANCED FEATURE: Risk Taking Propensity
        if (this.risk_taking_propensity > 0.7) {
            const trumpDefenses = coverMoves.filter(move => 
                move.cards && move.cards.some(card => card.suit === game.power_suit)
            );
            if (trumpDefenses.length > 0 && Math.random() < 0.4) {
                return this.getLowestCostDefense(trumpDefenses, game.power_suit);
            }
        }
        
        // Default: minimal defense
        return this.getLowestCostDefense(coverMoves, game.power_suit);
    }
    
    private selectPassMove(passMoves: LegalMove[], powerSuit: number): LegalMove {
        // Choose pass move with lowest value cards
        return passMoves.reduce((best, current) => {
            if (!current.cards || !best.cards) return current;
            
            const currentScore = this.getCardScore(current.cards, powerSuit);
            const bestScore = this.getCardScore(best.cards, powerSuit);
            
            return currentScore < bestScore ? current : best;
        });
    }
    
    private getBestAttackMove(attackMoves: LegalMove[], powerSuit: number): LegalMove {
        // Sort by attack preference (lower is better)
        return attackMoves.reduce((best, current) => {
            if (!current.cards || !best.cards) return current;
            
            const currentScore = this.getAttackScore(current.cards, powerSuit);
            const bestScore = this.getAttackScore(best.cards, powerSuit);
            
            return currentScore < bestScore ? current : best;
        });
    }
    
    private getLowestCostDefense(coverMoves: LegalMove[], powerSuit: number): LegalMove {
        return coverMoves.reduce((best, current) => {
            if (!current.cards || !best.cards) return current;
            
            const currentScore = this.getDefenseScore(current.cards, powerSuit);
            const bestScore = this.getDefenseScore(best.cards, powerSuit);
            
            return currentScore < bestScore ? current : best;
        });
    }
    
    private getAttackScore(cards: Card[], powerSuit: number): number {
        let score = 0;
        for (const card of cards) {
            if (card.suit === powerSuit) {
                score += card.value + 20;  // Trump cards are valuable for defense
            } else {
                score += card.value;
            }
        }
        return score;
    }
    
    private getDefenseScore(cards: Card[], powerSuit: number): number {
        let score = 0;
        for (const card of cards) {
            const isTrump = card.suit === powerSuit;
            score += card.value + (isTrump ? 10 : 0);
        }
        return score;
    }
    
    private getCardScore(cards: Card[], powerSuit: number): number {
        let score = 0;
        for (const card of cards) {
            score += card.value + (card.suit === powerSuit ? 20 : 0);
        }
        return score;
    }
    
    private shouldGiveUpBasic(hand: Card[], uncoveredAttacks: any[], game: Game): boolean {
        if (uncoveredAttacks.length === 0) return false;
        
        let totalTrumpCost = 0;
        let totalSameSuitCost = 0;
        let undefendableAttacks = 0;
        
        for (const battle of uncoveredAttacks) {
            const attack = battle.attack;
            const trumpOptions = hand.filter(card => 
                card.suit === game.power_suit && canCover(attack, card, game.power_suit)
            );
            const sameSuitOptions = hand.filter(card => 
                card.suit === attack.suit && card.value > attack.value
            );
            
            if (trumpOptions.length === 0 && sameSuitOptions.length === 0) {
                undefendableAttacks++;
            } else if (sameSuitOptions.length > 0) {
                // Can defend with same suit (preferred)
                totalSameSuitCost += Math.min(...sameSuitOptions.map(c => c.value));
            } else {
                // Must use trump
                totalTrumpCost += Math.min(...trumpOptions.map(c => c.value));
            }
        }
        
        // Give up if any attacks are undefendable
        if (undefendableAttacks > 0) {
            return true;
        }
        
        // Give up if defense would consume too many high-value cards
        if (uncoveredAttacks.length >= 3 && (totalTrumpCost + totalSameSuitCost) > 30) {
            return true;
        }
        
        // Give up if we'd have to use multiple high trumps
        const highTrumps = hand.filter(card => card.suit === game.power_suit && card.value >= 12);
        if (totalTrumpCost >= 24 && highTrumps.length >= 2) {
            return true;
        }
        
        return false;
    }
} 

File: cnitro/src/ultimate_champion_strategy.c
// Ultimate Champion strategy — direct port of UltimateChampionStrategy in
// supabase/functions/_shared/strategies/ultimate_champion_strategy.ts.
//
// The TS never uses Array.sort here — every "best move" selection is a
// `reduce` that only replaces the incumbent on a STRICT `<` of the score, so
// ties keep the earliest move in legalMoves order. We mirror that with
// single-pass first-strict-min scans over the (identically ordered) moves
// list; the TS `filter`s are folded into predicates, which preserves order.
//
// Math.random() maps to random_strategy_random(); the number and order of
// draws is preserved exactly (see the per-site comments below).
//
// The TS class also declares coverage_consideration / duplicate_preference /
// positional_strategy, but chooseMove never reads them — omitted.

#include "card.h"
#include "game.h"
#include "legal.h"

// ---------- tuned parameters (verbatim from the TS class fields) --------

static const bool   trump_count_awareness = true;   // +1.1% effect
static const bool   prefer_passing = true;          // +0.9% effect

static const double bluff_attack_prob = 0.25;
static const double give_up_threshold = 0.95;
static const double trump_conservation_level = 0.60;
static const double risk_taking_propensity = 0.40;
static const double opponent_hand_weight = 0.90;
static const double deck_size_sensitivity = 0.65;
static const double attack_continuation_prob = 0.75;
static const double defense_desperation_threshold = 0.45;
static const double passing_aggressiveness = 0.25;
static const double endgame_strategy_switch = 0.50;

// ---------- move predicates (the TS filter lambdas) ----------------------
// All of these are only applied to attack/cover/pass moves, whose `cards`
// array always exists and is non-empty, so the TS `move.cards && ...` guards
// are vacuously true and not mirrored.

// move.cards.some(card => card.value >= 10)
static bool pred_any_value_ge10(const LegalMove *m, int power_suit) {
    (void)power_suit;
    for (int i = 0; i < m->n_cards; i++) if (m->cards[i].value >= 10) return true;
    return false;
}
// move.cards.every(card => card.suit !== power_suit)
static bool pred_all_non_trump(const LegalMove *m, int power_suit) {
    for (int i = 0; i < m->n_cards; i++) if (m->cards[i].suit == power_suit) return false;
    return true;
}
// move.cards.some(card => card.suit === power_suit)
static bool pred_any_trump(const LegalMove *m, int power_suit) {
    for (int i = 0; i < m->n_cards; i++) if (m->cards[i].suit == power_suit) return true;
    return false;
}
// move.cards.some(card => card.value >= 8 && card.value <= 11)
static bool pred_any_value_8_to_11(const LegalMove *m, int power_suit) {
    (void)power_suit;
    for (int i = 0; i < m->n_cards; i++) {
        if (m->cards[i].value >= 8 && m->cards[i].value <= 11) return true;
    }
    return false;
}
// move.cards.every(card => card.value <= 8)
static bool pred_all_value_le8(const LegalMove *m, int power_suit) {
    (void)power_suit;
    for (int i = 0; i < m->n_cards; i++) if (m->cards[i].value > 8) return false;
    return true;
}

typedef bool (*MovePred)(const LegalMove *m, int power_suit);

// getAttackScore / getCardScore (trump_bonus = 20) and getDefenseScore
// (trump_bonus = 10) collapse to one sum.
static int score_cards(const LegalMove *m, int power_suit, int trump_bonus) {
    int score = 0;
    for (int i = 0; i < m->n_cards; i++) {
        score += m->cards[i].value + (m->cards[i].suit == power_suit ? trump_bonus : 0);
    }
    return score;
}

// The TS reduce pattern (getBestAttackMove / getLowestCostDefense /
// selectPassMove): first strict minimum over matching moves, in list order.
// Returns the move index, or -1 if no move matches (TS never reduces an
// empty filter result — callers check emptiness first).
static int reduce_lowest(const LegalMoves *moves, int type, MovePred pred,
                         int power_suit, int trump_bonus) {
    int best = -1;
    int best_score = 0;
    for (int i = 0; i < moves->n; i++) {
        const LegalMove *m = &moves->moves[i];
        if (m->type != type) continue;
        if (pred && !pred(m, power_suit)) continue;
        int s = score_cards(m, power_suit, trump_bonus);
        if (best < 0 || s < best_score) { best = i; best_score = s; }
    }
    return best;
}

static int count_matching(const LegalMoves *moves, int type, MovePred pred,
                          int power_suit) {
    int n = 0;
    for (int i = 0; i < moves->n; i++) {
        const LegalMove *m = &moves->moves[i];
        if (m->type == type && (!pred || pred(m, power_suit))) n++;
    }
    return n;
}

// k-th (0-based) matching move, mirroring filtered[Math.floor(rand * len)].
static int nth_matching(const LegalMoves *moves, int type, MovePred pred,
                        int power_suit, int k) {
    for (int i = 0; i < moves->n; i++) {
        const LegalMove *m = &moves->moves[i];
        if (m->type == type && (!pred || pred(m, power_suit))) {
            if (k == 0) return i;
            k--;
        }
    }
    return -1; // unreachable: k < count by construction
}

// Math.floor(Math.random() * n) for n >= 1: the draw is non-negative, so
// (int) truncation equals floor. Same IEEE ops as the TS.
static int random_index(int n) {
    int k = (int)(random_strategy_random() * (double)n);
    if (k >= n) k = n - 1; // unreachable guard (rand < 1.0)
    return k;
}

// ---------- shouldGiveUpBasic --------------------------------------------

static bool should_give_up_basic(const Game *g, const Player *bot) {
    // uncoveredAttacks = table_battles with defense === null.
    int uncovered = 0;
    for (int i = 0; i < g->num_battles; i++) {
        if (!g->table_battles[i].has_defense) uncovered++;
    }
    if (uncovered == 0) return false;

    int total_trump_cost = 0;
    int total_same_suit_cost = 0;
    int undefendable_attacks = 0;

    for (int i = 0; i < g->num_battles; i++) {
        if (g->table_battles[i].has_defense) continue;
        Card attack = g->table_battles[i].attack;

        // trumpOptions: trump cards that canCover; keep only the min value.
        bool has_trump_option = false;
        int min_trump = 0;
        // sameSuitOptions: same suit, strictly higher value; min value.
        bool has_same_suit = false;
        int min_same_suit = 0;
        for (int h = 0; h < bot->hand_count; h++) {
            Card c = bot->hand[h];
            if (c.suit == g->power_suit && can_cover(attack, c, g->power_suit)) {
                if (!has_trump_option || c.value < min_trump) min_trump = c.value;
                has_trump_option = true;
            }
            if (c.suit == attack.suit && c.value > attack.value) {
                if (!has_same_suit || c.value < min_same_suit) min_same_suit = c.value;
                has_same_suit = true;
            }
        }

        if (!has_trump_option && !has_same_suit) {
            undefendable_attacks++;
        } else if (has_same_suit) {
            total_same_suit_cost += min_same_suit; // can defend with same suit (preferred)
        } else {
            total_trump_cost += min_trump;         // must use trump
        }
    }

    // Give up if any attacks are undefendable.
    if (undefendable_attacks > 0) return true;

    // Give up if defense would consume too many high-value cards.
    if (uncovered >= 3 && (total_trump_cost + total_same_suit_cost) > 30) return true;

    // Give up if we'd have to use multiple high trumps.
    int high_trumps = 0;
    for (int h = 0; h < bot->hand_count; h++) {
        if (bot->hand[h].suit == g->power_suit && bot->hand[h].value >= 12) high_trumps++;
    }
    if (total_trump_cost >= 24 && high_trumps >= 2) return true;

    return false;
}

// ---------- ultimateGiveUpDecision ----------------------------------------

static bool ultimate_give_up_decision(const Game *g, const Player *bot) {
    bool base_should = should_give_up_basic(g, bot);

    // Give Up Threshold: one unconditional Math.random() draw.
    if (random_strategy_random() < (1.0 - give_up_threshold)) return true;

    // Deck Size Sensitivity (0.65 > 0.5: always active).
    if (deck_size_sensitivity > 0.5) {
        double deck_ratio = (double)g->deck_count / 36.0;
        if (deck_ratio > 0.5) {
            if (!base_should) return false;
        } else {
            if (base_should) return true;
        }
    }

    // Endgame Strategy Switch — dead in TS: 0.50 > 0.5 is false. Kept verbatim.
    if (bot->hand_count <= 3 && endgame_strategy_switch > 0.5) {
        int trump_cards = 0;
        for (int h = 0; h < bot->hand_count; h++) {
            if (bot->hand[h].suit == g->power_suit) trump_cards++;
        }
        if (trump_cards >= 2) return false;
    }

    return base_should;
}

// ---------- selectDefenseCard ---------------------------------------------

static int select_defense_card(const Game *g, const Player *bot,
                               const LegalMoves *moves) {
    // Trump Count Awareness (hard-coded true).
    if (trump_count_awareness) {
        int trump_count = 0;
        for (int h = 0; h < bot->hand_count; h++) {
            if (bot->hand[h].suit == g->power_suit) trump_count++;
        }
        double conservation_multiplier = 1.0 + trump_conservation_level;
        // Math.floor on positive doubles == (int) truncation.
        int many_trumps_threshold = (int)(4.0 * conservation_multiplier);
        int few_trumps_threshold = (int)(2.0 / conservation_multiplier);
        if (few_trumps_threshold < 1) few_trumps_threshold = 1; // Math.max(1, ...)

        if (trump_count >= many_trumps_threshold) {
            int idx = reduce_lowest(moves, MOVE_COVER, pred_any_trump, g->power_suit, 10);
            if (idx >= 0) return idx;
        }
        if (trump_count <= few_trumps_threshold) {
            int idx = reduce_lowest(moves, MOVE_COVER, pred_all_non_trump, g->power_suit, 10);
            if (idx >= 0) return idx;
        }
    }

    // Defense Desperation Threshold: one draw, only when hand is small.
    if (bot->hand_count <= 3) {
        double desperation_level = 1.0 - defense_desperation_threshold;
        if (random_strategy_random() < desperation_level) {
            int idx = reduce_lowest(moves, MOVE_COVER, pred_any_value_ge10, g->power_suit, 10);
            if (idx >= 0) return idx;
        }
    }

    // Risk Taking Propensity — dead in TS: 0.40 > 0.7 is false. Kept verbatim.
    // (The TS only draws Math.random() when trumpDefenses is non-empty.)
    if (risk_taking_propensity > 0.7) {
        int idx = reduce_lowest(moves, MOVE_COVER, pred_any_trump, g->power_suit, 10);
        if (idx >= 0 && random_strategy_random() < 0.4) return idx;
    }

    // Default: minimal defense.
    return reduce_lowest(moves, MOVE_COVER, NULL, g->power_suit, 10);
}

// ---------- selectDefenseMove ----------------------------------------------

static int select_defense_move(const Game *g, int bot_idx, const LegalMoves *moves) {
    const Player *bot = &g->players[bot_idx];

    int uncovered = 0;
    for (int i = 0; i < g->num_battles; i++) {
        if (!g->table_battles[i].has_defense) uncovered++;
    }
    if (uncovered > 0) {
        if (ultimate_give_up_decision(g, bot)) {
            for (int i = 0; i < moves->n; i++) {
                if (moves->moves[i].type == MOVE_PICKUP) return i; // strategic pickup
            }
        }
    }

    int idx = select_defense_card(g, bot, moves);
    if (idx >= 0) return idx;

    return 0; // default fallback: legalMoves[0]
}

// ---------- selectAttackMove ------------------------------------------------

static int select_attack_move(const Game *g, const LegalMoves *moves) {
    // (The TS also reads the defender's hand size here but never uses it.)

    // Bluff Attack Probability: one unconditional draw; a second draw only
    // when the bluff fires AND a >=10 attack exists.
    if (random_strategy_random() < bluff_attack_prob) {
        int n = count_matching(moves, MOVE_ATTACK, pred_any_value_ge10, g->power_suit);
        if (n > 0) {
            return nth_matching(moves, MOVE_ATTACK, pred_any_value_ge10,
                                g->power_suit, random_index(n));
        }
    }

    // Trump Conservation Level (0.60 > 0.5: always active).
    if (trump_conservation_level > 0.5) {
        int idx = reduce_lowest(moves, MOVE_ATTACK, pred_all_non_trump, g->power_suit, 20);
        if (idx >= 0) return idx;
    }

    // Risk Taking Propensity — both arms dead in TS (0.40). Kept verbatim.
    if (risk_taking_propensity > 0.7) {
        int n = count_matching(moves, MOVE_ATTACK, pred_any_value_8_to_11, g->power_suit);
        if (n > 0) {
            return nth_matching(moves, MOVE_ATTACK, pred_any_value_8_to_11,
                                g->power_suit, random_index(n));
        }
    } else if (risk_taking_propensity < 0.3) {
        int idx = reduce_lowest(moves, MOVE_ATTACK, pred_all_value_le8, g->power_suit, 20);
        if (idx >= 0) return idx;
    }

    // Default: smart attack preference.
    return reduce_lowest(moves, MOVE_ATTACK, NULL, g->power_suit, 20);
}

// ---------- chooseMove --------------------------------------------------------

int ultimate_champion_strategy_choose(const Game *g, int bot_idx,
                                      const LegalMoves *moves, void *ctx) {
    (void)ctx;
    if (moves->n == 0) return -1; // TS throws

    bool is_defender = (bot_idx == g->defender);
    bool is_attacker = (bot_idx == g->first_attacker) || !is_defender;

    // One scan for the filters the TS builds up front / lazily.
    int first_good = -1, first_pass = -1, first_wait = -1, first_pickup = -1;
    bool has_attack = false, has_cover = false;
    for (int i = 0; i < moves->n; i++) {
        switch (moves->moves[i].type) {
        case MOVE_GOOD:   if (first_good < 0) first_good = i;     break;
        case MOVE_ATTACK: has_attack = true;                      break;
        case MOVE_COVER:  has_cover = true;                       break;
        case MOVE_PASS:   if (first_pass < 0) first_pass = i;     break;
        case MOVE_WAIT:   if (first_wait < 0) first_wait = i;     break;
        case MOVE_PICKUP: if (first_pickup < 0) first_pickup = i; break;
        default: break;
        }
    }

    // Attack Continuation Probability via "good": one draw when a good move
    // exists and there is at least one battle on the table.
    if (first_good >= 0 && g->num_battles >= 1) {
        double continue_prob = attack_continuation_prob;
        int defender_hand_size = g->players[g->defender].hand_count;

        // Opponent Hand Size Weight (0.90 > 0.5: always active).
        if (opponent_hand_weight > 0.5) {
            if (defender_hand_size <= 2) {
                continue_prob *= 1.5; // more pressure against weak opponents
            } else if (defender_hand_size >= 5) {
                continue_prob *= 0.7; // less pressure against strong opponents
            }
        }

        if (random_strategy_random() > continue_prob) return first_good;
    }

    // Attack strategy.
    if (has_attack && is_attacker) return select_attack_move(g, moves);

    // Defense strategy.
    if (has_cover && is_defender) return select_defense_move(g, bot_idx, moves);

    // Pass strategy (prefer_passing hard-coded true).
    if (prefer_passing) {
        if (first_pass >= 0) {
            // Passing Aggressiveness: one draw when a pass move exists.
            double pass_chance = 0.3 + (passing_aggressiveness * 0.4);
            if (random_strategy_random() < pass_chance) {
                // TS: (game.defender + 1) % game.players.length — includes
                // OUT players; the `if (nextDefender && ...)` existence check
                // is always true since the index is in range.
                int next_defender = (g->defender + 1) % g->num_players;
                if (g->players[next_defender].hand_count
                        <= g->players[bot_idx].hand_count) {
                    // selectPassMove: lowest getCardScore (trump +20).
                    return reduce_lowest(moves, MOVE_PASS, NULL, g->power_suit, 20);
                }
            }
        }
    }

    // Good moves (if not chosen earlier probabilistically).
    if (first_good >= 0) return first_good;

    // Wait when appropriate (TS also console.logs here — side effect only).
    if (first_wait >= 0) return first_wait;

    // Pickup as last resort.
    if (first_pickup >= 0) return first_pickup;

    // Default: first available move.
    return 0;
}
